File: frontend/src/lib/stores/cues.ts
/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import { writable, get } from 'svelte/store';
import * as Commands from '../../../bindings/github.com/stexxo/dynocue/internal/gui/commands';
import { Cue } from '../../../bindings/github.com/stexxo/dynocue/api/cues';
import { Events } from '@wailsio/runtime';

function createCueStore() {
    const lists = new Map<number, {
        subscribe: (run: (value: Cue[]) => void) => () => void;
        set: (value: Cue[]) => void;
        update: (updater: (value: Cue[]) => Cue[]) => void;
    }>();

    function getStore(cueListNumber: number) {
        let store = lists.get(cueListNumber);
        if (!store) {
            store = writable<Cue[]>([]);
            lists.set(cueListNumber, store);
        }
        return store;
    }

    async function refresh(cueListNumber: number) {
        const store = getStore(cueListNumber);
        try {
            const result = await Commands.EnumerateCue({ cueListNumber });
            if (result && result.cues) {
                store.set(result.cues.map((c: any) => c.cue));
            } else {
                store.set([]);
            }
        } catch (err) {
            console.error(`Failed to enumerate cues for cue list ${cueListNumber}:`, err);
        }
    }

    async function updateMetadata(cueListNumber: number, cueNumber: number, key: string, value: string) {
        try {
            await Commands.UpdateCue({ cueListNumber, cueNumber, key, value });
        } catch (err) {
            console.error(`Failed to update cue ${cueNumber}:`, err);
        }
    }

    async function create(cueListNumber: number, cueNumber: number = 0) {
        try {
            await Commands.CreateCue({ cueListNumber, cueNumber });
        } catch (err) {
            console.error('Failed to create cue:', err);
        }
    }

    async function remove(cueListNumber: number, cueNumber: number) {
        try {
            await Commands.DeleteCue({ cueListNumber, cueNumber });
        } catch (err) {
            console.error(`Failed to delete cue ${cueNumber}:`, err);
        }
    }

    async function move(cueListNumber: number, originalCueNumber: number, newCueNumber: number) {
        try {
            await Commands.MoveCue({ cueListNumber, originalCueNumber, newCueNumber });
        } catch (err) {
            console.error(`Failed to move cue ${originalCueNumber} to ${newCueNumber}:`, err);
        }
    }

    // Subscribe to backend events
    Events.On('event.cue.created', (event: any) => {
        const { cueListNumber, cue } = event.data;
        const store = lists.get(cueListNumber);
        if (!store) return;

        store.update(cues => {
            const newCues = [...cues, cue];
            return newCues.sort((a, b) => a.cueNumber - b.cueNumber);
        });
    });

    Events.On('event.cue.updated', (event: any) => {
        const { cueListNumber, cue } = event.data;
        const store = lists.get(cueListNumber);
        if (!store) return;

        store.update(cues => cues.map(c => 
            c.cueNumber === cue.cueNumber ? cue : c
        ));
    });

    Events.On('event.cue.deleted', (event: any) => {
        const { cueListNumber, cueNumber } = event.data;
        const store = lists.get(cueListNumber);
        if (!store) return;

        store.update(cues => cues.filter(cue => cue.cueNumber !== cueNumber));
    });

    return {
        byListNumber: (cueListNumber: number) => {
            const store = getStore(cueListNumber);
            return {
                subscribe: store.subscribe,
                refresh: () => refresh(cueListNumber),
                updateMetadata: (cueNumber: number, key: string, value: string) => updateMetadata(cueListNumber, cueNumber, key, value),
                create: (cueNumber: number = 0) => create(cueListNumber, cueNumber),
                remove: (cueNumber: number) => remove(cueListNumber, cueNumber),
                move: (originalCueNumber: number, newCueNumber: number) => move(cueListNumber, originalCueNumber, newCueNumber)
            };
        }
    };
}

export const cues = createCueStore();

File: internal/core/cues/cues.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package cues

import (
	"errors"
	"log/slog"
	"slices"

	apicues "gitlab.com/stexxo/dynocue/api/cues"
	"gitlab.com/stexxo/dynocue/internal/data"
	"gitlab.com/stexxo/dynocue/internal/utils"
	apibus "gitlab.com/stexxo/dynocue/pkg/bus"
	"go.etcd.io/bbolt"
	berrors "go.etcd.io/bbolt/errors"
)

// NewCue creates a new cue within a specific cue list.
func (c *CueSystem) NewCue(sub string, in apicues.CreateCueInput) (*apibus.MessageResponse[apicues.CreateCueOutput], error) {
	md := &CueMetadataDbModel{}
	var outNum float64
	err := c.db.Update(func(tx *bbolt.Tx) error {
		_, key, err := data.AddIncrementedSubBucket(
			tx,
			BucketCueListKey,
			[]data.BucketKey{data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey},
			in.CueNumber,
			data.NewResourceBootstrap{
				InitialValues: []data.KeyValuePair{{Key: KeyMetadata, Value: md}},
				Buckets:       []data.BucketKey{BucketActionsKey},
			},
		)
		outNum = key
		return err
	})
	if err != nil {
		slog.Error("failed to create new cue", "err", err.Error(), "number", outNum, "cueList", in.CueListNumber)
		return apibus.NewMessageResponse[apicues.CreateCueOutput](nil, apibus.NewMessageError("failed to create cue")), nil
	}

	if err = apibus.Publish(c.conn, apicues.EventNewCue, apicues.NewCueEvent{
		CueListNumber: in.CueListNumber,
		Cue: apicues.Cue{
			CueNumber:   outNum,
			Label:       md.Label,
			Description: md.Description,
		},
	}); err != nil {
		slog.Error("failed to publish change event for new cue", slog.String("err", err.Error()))
	}

	return &apibus.MessageResponse[apicues.CreateCueOutput]{
		ResponseValue: &apicues.CreateCueOutput{
			CueListNumber: in.CueListNumber,
			CueNumber:     outNum,
		},
	}, nil
}

// UpdateCue updates the metadata fields of an existing cue.
func (c *CueSystem) UpdateCue(sub string, in apicues.UpdateCueInput) (*apibus.MessageResponse[apicues.UpdateCueOutput], error) {
	var outMetadata *CueMetadataDbModel
	err := c.db.Update(func(tx *bbolt.Tx) error {
		b, err := data.GetBucketFromRoot(tx, false, BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey, data.NewFloatBucketKey(in.CueNumber, false))
		if err != nil {
			return err
		}
		outMetadata, err = data.UpdateAttributeInKeyValuePair[CueMetadataDbModel](b, KeyMetadata, in.Key, in.Value)
		return err
	})
	if err != nil {
		slog.Error("failed to update cue", "err", err.Error(), "cue", in.CueNumber, "cueList", in.CueListNumber)
		return apibus.NewMessageResponse[apicues.UpdateCueOutput](nil, apibus.NewMessageError("failed to update cue")), nil
	}

	if err = apibus.Publish(c.conn, apicues.EventUpdateCue, apicues.UpdateCueEvent{
		CueListNumber: in.CueListNumber,
		Cue: apicues.Cue{
			CueNumber:   in.CueNumber,
			Label:       outMetadata.Label,
			Description: outMetadata.Description,
		},
	}); err != nil {
		slog.Error("failed to publish change event for update cue", slog.String("err", err.Error()))
	}

	return &apibus.MessageResponse[apicues.UpdateCueOutput]{
		ResponseValue: &apicues.UpdateCueOutput{},
	}, nil
}

// GetCue retrieves a specific cue.
func (c *CueSystem) GetCue(sub string, in apicues.GetCueInput) (*apibus.MessageResponse[apicues.GetCueOutput], error) {
	var md *CueMetadataDbModel
	err := c.db.View(func(tx *bbolt.Tx) error {
		b, err := data.GetBucketFromRoot(tx, true, BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey, data.NewFloatBucketKey(in.CueNumber, false))
		if err != nil {
			return err
		}
		md = new(CueMetadataDbModel)
		return data.GetKey(b, KeyMetadata, md)
	})

	if err != nil {
		slog.Error("failed to get cue", "err", err.Error(), "cue", in.CueNumber, "cueList", in.CueListNumber)
		return apibus.NewMessageResponse[apicues.GetCueOutput](nil, apibus.NewMessageError("failed to retrieve cue")), nil
	}

	return &apibus.MessageResponse[apicues.GetCueOutput]{
		ResponseValue: &apicues.GetCueOutput{
			CueListNumber: in.CueListNumber,
			Cue: apicues.Cue{
				CueNumber:   in.CueNumber,
				Label:       md.Label,
				Description: md.Description,
			},
		},
	}, nil
}

// EnumerateCue returns a list of all cues within a specific cue list.
func (c *CueSystem) EnumerateCue(sub string, in apicues.EnumerateCueInput) (*apibus.MessageResponse[apicues.EnumerateCueOutput], error) {
	values := map[float64]*CueMetadataDbModel{}
	err := c.db.View(func(tx *bbolt.Tx) error {
		b, err := data.GetBucketFromRoot(tx, true, BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey)
		if errors.Is(err, data.ErrNoBucket) {
			return nil
		}
		if err != nil {
			return err
		}
		values, err = data.EnumerateKeysFromSubBuckets[float64, CueMetadataDbModel](b, KeyMetadata, func(bytes []byte) float64 {
			k, err := utils.BytesToFloat64(bytes)
			if err != nil {
				return 0
			}
			return k
		})
		return err
	})
	if err != nil {
		slog.Error("failed to enumerate cues", "err", err.Error(), "cuelist", in.CueListNumber)
		return apibus.NewMessageResponse[apicues.EnumerateCueOutput](nil, apibus.NewMessageError("failed to enumerate cues")), nil
	}

	out := make([]apicues.GetCueOutput, 0, len(values))
	for k, v := range values {
		i, _ := slices.BinarySearchFunc(out, k, func(a apicues.GetCueOutput, b float64) int {
			if a.Cue.CueNumber < b {
				return -1
			} else if a.Cue.CueNumber > b {
				return 1
			}
			return 0
		})
		out = slices.Insert(out, i, apicues.GetCueOutput{
			CueListNumber: in.CueListNumber,
			Cue: apicues.Cue{
				CueNumber:   k,
				Label:       v.Label,
				Description: v.Description,
			},
		})
	}

	return &apibus.MessageResponse[apicues.EnumerateCueOutput]{
		ResponseValue: &apicues.EnumerateCueOutput{
			CueListNumber: in.CueListNumber,
			Cues:          out,
		},
	}, nil
}

// DeleteCue removes a specific cue from a cue list.
func (c *CueSystem) DeleteCue(sub string, in apicues.DeleteCueInput) (*apibus.MessageResponse[apicues.DeleteCueOutput], error) {

	err := c.db.Update(func(tx *bbolt.Tx) error {
		return data.DeleteBucketByPath(tx, utils.Float64ToBytes(in.CueNumber), BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey)
	})

	if err != nil && !errors.Is(err, berrors.ErrBucketNotFound) {
		slog.Error("failed to delete cue", "err", err.Error(), "cueList", in.CueListNumber, "cue", in.CueNumber)
		return apibus.NewMessageResponse[apicues.DeleteCueOutput](nil, apibus.NewMessageError("failed to delete cue")), nil
	}

	if err = apibus.Publish(c.conn, apicues.EventDeleteCue, apicues.DeleteCueEvent{
		CueListNumber: in.CueListNumber,
		CueNumber:     in.CueNumber,
	}); err != nil {
		slog.Error("failed to publish change event for delete cue", slog.String("err", err.Error()))
	}

	return &apibus.MessageResponse[apicues.DeleteCueOutput]{
		ResponseValue: &apicues.DeleteCueOutput{},
	}, nil
}

// MoveCue changes the number of an existing cue.
func (c *CueSystem) MoveCue(sub string, in apicues.MoveCueInput) (*apibus.MessageResponse[apicues.MoveCueOutput], error) {
	if in.OriginalCueNumber == in.NewCueNumber {
		return &apibus.MessageResponse[apicues.MoveCueOutput]{
			ResponseValue: &apicues.MoveCueOutput{
				CueListNumber: in.CueListNumber,
				NewCueNumber:  in.NewCueNumber,
			},
		}, nil
	}

	err := c.db.Update(func(tx *bbolt.Tx) error {
		b, err := data.GetBucketFromRoot(tx, false, BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey)
		if err != nil {
			return err
		}

		err = data.RenameSubBucket(b, in.OriginalCueNumber, in.NewCueNumber)
		return err
	})

	if err != nil {
		slog.Error("failed to move cue", "err", err.Error(), "cueList", in.CueListNumber, "cue", in.OriginalCueNumber, "newCueNumber", in.NewCueNumber)
		return apibus.NewMessageResponse[apicues.MoveCueOutput](nil, apibus.NewMessageError("failed to move cue")), nil
	}

	outMetadata := &CueMetadataDbModel{}
	err = c.db.View(func(tx *bbolt.Tx) error {
		b, err := data.GetBucketFromRoot(tx, true, BucketCueListKey, data.NewFloatBucketKey(in.CueListNumber, false), BucketCuesKey, data.NewFloatBucketKey(in.NewCueNumber, false))
		if err != nil {
			return err
		}

		err = data.GetKey(b, KeyMetadata, outMetadata)
		return err
	})
	if err != nil {
		slog.Error("failed to get data about cue after move", "err", err.Error(), "cueList", in.CueListNumber, "originalCueNumber", in.OriginalCueNumber, "newCueNumber", in.NewCueNumber)
		return apibus.NewMessageResponse[apicues.MoveCueOutput](nil, apibus.NewMessageError("failed to get data about cue after move")), nil
	}

	// Emit events: delete old, create new
	if err = apibus.Publish(c.conn, apicues.EventDeleteCue, apicues.DeleteCueEvent{
		CueListNumber: in.CueListNumber,
		CueNumber:     in.OriginalCueNumber,
	}); err != nil {
		slog.Error("failed to publish delete event for move cue", slog.String("err", err.Error()))
	}

	if err = apibus.Publish(c.conn, apicues.EventNewCue, apicues.NewCueEvent{
		CueListNumber: in.CueListNumber,
		Cue: apicues.Cue{
			CueNumber: in.NewCueNumber,
			Label:     outMetadata.Label,
		},
	}); err != nil {
		slog.Error("failed to publish create event for move cue", slog.String("err", err.Error()))
	}

	return &apibus.MessageResponse[apicues.MoveCueOutput]{
		ResponseValue: &apicues.MoveCueOutput{
			CueListNumber: in.CueListNumber,
			NewCueNumber:  in.NewCueNumber,
		},
	}, nil
}
